### PR TITLE
Replace more uses of std::enable_if for SFINAE by concepts

### PR DESCRIPTION
**Pull request by rschu1ze**
- enable_if is usually regarded as fragile and unreadable
- C++20 concepts are much easier to read and produce more expressive error messages

- this is follow-up to PR #35347 but this time, most of the remaining and more complex usages of enable_if in the codebase were replaced

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by kitaisreal** [comment]
@Mergifyio update

File: src/Common/HashTable/HashTable.h [hunk 1]
```diff
@@ -1036,14 +1036,16 @@ class HashTable :
         return const_cast<std::decay_t<decltype(*this)> *>(this)->find(x, hash_value);
     }
 
-    std::enable_if_t<Grower::performs_linear_probing_with_single_step, bool>
-    ALWAYS_INLINE erase(const Key & x)
+    template <typename Grower_ = Grower>
+    requires (Grower_::performs_linear_probing_with_single_step)
```
Comment (kitaisreal) [inline src/Common/HashTable/HashTable.h:1040]: Here it is better to leave previous implementation. With concepts it does not look obvious. Same for function below.

**Review by kitaisreal — CHANGES_REQUESTED** [review]
Pull request looks good. Need to fix HashMap concept usage and ready to merge.

File: src/Common/HashTable/HashTable.h (hunk 1 above)
Comment (rschu1ze) [reply src/Common/HashTable/HashTable.h:1040]: I agree, and will provide a fix as requested.

**Review by kitaisreal — APPROVED** [review]

**Pull request merged**